spec/immutable-data-structures.html

```diff
@@ -16,8 +16,20 @@ <h1>The Record Constructor</h1>
         <li>is not intended to be used with the `new` operator or to be subclassed. It may be used as the value of an *extends* clause of a class definition but a *super* call to the Record constructor will cause an exception.</li>
       </ul>
       <emu-clause id="sec-record-constructor-record-value">
-        <h1>Record ( _value_ )</h1>
-        <p>TODO</p>
+        <h1>Record ( _arg_ )</h1>
+        <p>When the `Record` function is called, the following steps are taken:</p>
+        <emu-alg>
+          1. If NewTarget is not *undefined*, throw a *TypeError* exception.
+          1. Let _obj_ be ? ToObject(_arg_).
+          1. Let _nameList_ be ? EnumerableOwnPropertyNames(_obj_, ~key~).
+          1. Let _fields_ be a new empty List.
+          1. For each element _name_ of _nameList_,
+            1. Let _value_ be ? GetV(_obj_, _name_).
+            1. If Type(_value_) is Object, throw a *TypeError* exception.
+            1. Let _field_ be { [[Key]]: _name_, [[Value]]: _value_ }.

```

Comment:
is it intentional that this will throw an error (observably via getters) one item at a time? If you pass `~key+value~` then all the value getting will be done up front.
```suggestion
          1. Let _nameList_ be ? EnumerableOwnPropertyNames(_obj_, ~key+value~).
          1. Let _fields_ be a new empty List.
          1. For each element _entry_ of _nameList_,
            1. Let _name_ be ! Get(_entry_, `"0"`).
            1. Let _value_ be ! Get(_entry_, `"1"`).
            1. If Type(_value_) is Object, throw a *TypeError* exception.
            1. Let _field_ be { [[Key]]: _name_, [[Value]]: _value_ }.
```

Reply:
It was intentional, although an argument could be made either way. @littledan suggested that I check/throw in each iteration of `Tuple.prototype.map`, and I figured that logic should apply here as well.

In general, what is the approach for things like this? Is there prior art or reasoning for interleaving the "gets" and type checks vs saving type checks for the end?

Reply:
I don't know if there's a strong precedent either way; when using an iterable, throwing midway seems preferable, but for objects, i could see it going either way. To me it doesn't make much sense to throw midway through iterating an object, since an object isn't an iterable, it's a "bag of stuff".

Reply:
Seems reasonable to me. I switched to using `key+value`.